docs/language-specification.md

```diff
@@ -0,0 +1,204 @@
+[P4 language]: https://p4.org/specs/
+[P4 language specification]: https://p4.org/specs/
+
+## Tour of the language
+
+The p4-constraints language is a simple, expression-based language for
+specifying properties of P4 table entries.
+
+### Keys and comparisons
+
+At the heart of the language lies the ability to talk about table keys,
+and to constrain their values using comparison operators.
+```p4
+@entry_restriction("ipv4.valid == 1")
+table ipv4_table {
+  key = {
+    ipv4.isValid() : exact @name("ipv4.valid");
+    ipv4.dst : ternary;
+  }
+}
+```
+For example, the above constraint states that the key `ipv4.valid` must be equal
+to `1` for all table entries. Just like P4, the constraint language provides the
+comparison operators `==`, `!=`, `>`, `>=`, `<`, and `<=` with the usual
+semantics.
+
+### Accessing the fields of a key
+
+While an `exact` match consists of a single value, a `ternary` match is given
+by a value together with a mask. Similarly, an `lpm` match consists of a value
+together with a prefix length, and a range match consists of lower and upper
+bounds.
+
+To access the different values associated with a key, the language provides a
+field access (or "projection") operator `::`.
+```p4
+@entry_restriction("ipv6.dst::prefix_length <= 64")
+table ipv6_table {
+  key = {
+    ipv6.dst : lpm;
+  }
+}
+```
+For example, the above constraint states that entries should only ever match on
+the most-significant 64 bits of an IPv6 address.
+
+More generally, assume we have a P4 table of the form
+```p4
+table t {
+  key = {
+    k : <match_type>;
+  }
+}
+```
+with `k` being of type `bit<W>`. Then the fields that can be accessed using
+the `::` operator together with their types are summarized in the following
+table:
+
+| <match_type> | field            | field type |
+|--------------|------------------|------------|
+| exact        | k::value         | bit\<W\>   |
+| ternary      | k::value         | bit\<W\>   |
+|              | k::mask          | bit\<W\>   |
+| lpm          | k::value         | bit\<W\>   |
+|              | k::prefix_length | int        |
+| range        | k::low           | bit\<W\>   |
+|              | k::high          | bit\<W\>   |
+
+When `k` is of type `bool`, everything behaves precisely as if `k` was of type
+`bit<1>`, with the boolean constant `true` and `false` being mapped to `1` and
+`0`, respectively.
+
+### Implicit conversions
+TODO
+
+### Boolean operations
+
+As in P4, Boolean expressions can be negated using `!`,
+or combined conjunctively using `&&` or disjunctively using `||`.
+
+Additionally, p4-constraints provides the binary operator `->` to denote
+[logical implication](https://en.wikipedia.org/wiki/Material_conditional):
+```p4
+@entry_restriction("ipv4.dst::mask != 0 -> ipv4.valid == 1")
+table ipv4_table {
+  key {
+    ipv4.isValid() : exact @name("ipv4.valid");
+    ipv4.dst : ternary;
+  }
+}
+```
+For example, the entry restriction above demands that non-trivial matches on
```

Comment:
Can we say what "non-trivial matches" is? I know what it is, but might be worth spelling it out.

Reply:
Might also be worth saying how that's encoded using a implication. Obvious for us, but maybe less so for the average programmer?

Reply:
Let me know if this is better.